### DISCUSSION TITLE
Assembly details and versioning extension (from user projects)

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [OrchardCMS/OrchardCore#11297] [Ideas]
Apologies if the verbiage is imprecise. Is there a _widget_ or what have you to convey top level, presumably _Orchard CMS_ enabled app, assembly details, copyright, versions, or just in general `AssemblyName` based details, etc? If not how difficult would that be perhaps? Thanks!

**Comment by mwpowellhtx** [comment]
This is just me thinking out loud, could possibly start by extending `OrchardCore.ContentManagement.Metadata.Models.ContentPartDefinition` and/or some `OrchardCore.ContentManagement.Metadata.Models.ContentPartFieldDefinition`. Based on a loose acquaintance with the `OrchardCore` namespace. Perhaps with some `Startup` _configuration_ extension methods that allow connection with an `Assembly`, `AssemblyName`, etc.

Anyone's thoughts 💡 ?

Thank you...

**Reply by mwpowellhtx** [reply]
Okay, so getting brave, still with questions on the back end... Would need to arrange that in some content types or what not to get a better idea apples from oranges there.

![image](https://user-images.githubusercontent.com/1824187/156841692-b53eadde-01bc-4d44-beca-76ff6fbd6e7f.png)

**Comment by mwpowellhtx** [comment]
To summarize, what I am after is an easy-to-(re-)use widget, `ContentType` (?), _item_, that I can inject for a given site.

**Reply by Skrypt** [reply]
You seem to talk about using a custom module in that case. You can use `dotnet new` to create modules or themes provided with Orchard Core as easy startup modules. Please look at the documentation for this. And yes, that would mean that you would need to extend some classes in that module to get some reusable ContentParts.

**Reply by mwpowellhtx** [reply]
Okay; so `ContentParts` is the starting point? That's pretty much what I'm interested in, key starting points there. Probably then 2+ modules, for the `Assembly` credentials, another for theme(s), etc. I can fumble around from there, probably. Thanks...

**Reply by mwpowellhtx** [reply]
Not sure I quite understand the docs; however, starting from an _Orchard Core Cms Module_ project template_. I setup a module, pretty sure I've named at least a stub for my `ContentPart`, i.e.

```csharp
public class AssemblyCredentialsPart : ContentPart
{
    // ...
}
```

Also _controllers_, _drivers_, _handlers_, and so forth.

I can fill in some properties there? i.e. an `Assembly` and/or `AssemblyName` property? Which I asssume are somehow relayed to the _CMS_ _admin_ _configuration_ and so forth?

How is that connected via a web app, one `Startup` to the module `Startup`? Let alone identifying requisite bits, `Assembly`, etc, to the parts.

**Reply by mwpowellhtx** [reply]
Okay, so far, seems to connect the dots... TBD in terms of arranging with a _content type_ (?) or what not.

```csharp
public void ConfigureServices(IServiceCollection services)
{
    // ...
    services.AddOrchardCms(config =>
    {
        // ...
        config.RegisterStartup<Path.To.My.Cms.Module.Startup>();
        // ...
    });
    // ...
}
```

**Comment by mwpowellhtx** [comment]
Maybe a follow on question here, `OrchardCore.Modules.Manifest.ModuleAttribute` ... what is it.

Eventually, I want to connect a `.csproj` build properties bump to assembly versions and so forth, i.e. bumps during build under prescribed conditions, and I want to relay that to `ModuleAttribute`, possibly. Possible? If so, how? Big question there would be _when_ the attribute is resolved perhaps as a task with dependency on the bump happening first?

**Reply by mwpowellhtx** [reply]
Along same lines, as part of the `.csproj` rather than a `Manifest.cs`, per se. Seems that could potentially be far more appropriate IMO.

**Comment by mwpowellhtx** [comment]
Re: `Migrations`, what are they to a _parts module_? Not sure I quite understand that. Are we specifying a JSON translation? What happens as a part potentially changes from one _version_ or _commit_ to the next, we need to do another migration? For starters... Thanks!

**Reply by mwpowellhtx** [reply]
Just observing, I sort of get it.

```csharp
// A OrchardCore.Data.Migration.DataMigration is discovered through method reflection
// and runs sequentially from the Create() method through UpdateFromX().
```

So if `Create()` is truly required then for this to take place, then should that not be specified. For that matter, with descriptive comments, especially on the expected return value, assumptions, etc, i.e. sorted how? In ascending order? Logically, perhaps, but dare we assume?

```csharp
/// <summary>
/// ...
/// </summary>
/// <returns>With more detail in the Create return value.</returns>
public abstract int Create();
```

**Reply by Skrypt** [reply]
Create() is executed when you enable a feature in Orchard Core. Then, it adds a record in the database about which version of migration has been executed last, so in that case, it is the returned value from the Create method. Though, if you want to upgrade the module later on you can add other methods in there so that it executes on top of already running tenants that have this module enabled.

**Reply by mwpowellhtx** [reply]
> you can add other methods in there...

Huh? That is not how I read it from the base class comments, but I could be wrong. I read to mean it is more of an interface in the style and manner of its behavior. i.e. _Create_ is specifying definitions and such through the man in the middle pattern i.e. `_contentDefinitionManager`, and a _'version'_, after a sort, though somewhat crude, is sorted and applied as such. i.e. if you updated version 3 and had 4-6 versions pending, those would be applied next time round. More clarity there would be fantastic, specify an interface, even an abstract method with comments.

**Reply by mwpowellhtx** [reply]
So in other words, I read to mean migrations should be approached more like this:

```csharp
public class Migration1 : DataMigration
{
  public int Create() => 1;
}
public class Migration2 : DataMigration
{
  public int Create() => 2;
}
public class Migration3 : DataMigration
{
  public int Create() => 3;
}
```

With, presumably, `DataMigration` derivations being consumed by convention.

**Reply by Skrypt** [reply]
The create method is generally used to define the base data structure for someone starting with a fresh install. Then the other methods ... `UpdateFrom1() { return 2; }` ... would be applied only for those who don't have the latest database structure or even data properly set for the module to work properly. So, this is where we run DB data and schema migrations. This is quite similar to something done in EntityFramework as an example.

**Reply by mwpowellhtx** [reply]
Ah, okay. Curious, and interesting. Seems unnecessarily fragile to me, IMO. From previous lives and experience, generally, I just have a migration class, somehow specify a _'version'_, i.e. literally a `System.Version` instance, let's say, and order by that, and simply apply the ones in balance. That's all migrations are as far as I'm concerned. And with reflection and such, it is super simply to identify, create, and order instances.

**Reply by mwpowellhtx** [reply]
@Skrypt Remember also from a principled perspective, [SOLID](https://simple.wikipedia.org/wiki/SOLID_(object-oriented_design)) is key IMO. Single instance single responsibility; one class responsible for one migration only; and then, depending on the migration pattern, _up/down/side grades_ as appropriate as well.

**Comment by mwpowellhtx** [comment]
Re: _Views_ and such, it is sort of apparent what is going on there. Not sure however, what is the `HomeController` about? Or the corresponding `Home/Index.cshtml` view? Supporting _settings_, _editing_, etc, purposes, via the _CMS_?

**Reply by Skrypt** [reply]
The Home page in Orchard Core is a page that can be set through the AutoroutePart settings. On a specific Content Type, you can set the AutoroutePart to enable setting the current content item to be the home page. HomeController in our demos or examples is just for that matter. It could have been named TestController too.

**Reply by mwpowellhtx** [reply]
Cool, sort of understood. In which context, the configuration, editor? Or live rendered content? Or both...

**Reply by Skrypt** [reply]
Both, the home page is a tenant setting altered by that AutoroutePart. And of course live or from recipe.

**Comment by mwpowellhtx** [comment]
Okay, so one or two steps back and gaining perspective... I'm not sure there is anything I need/want to convey re: _settings_ in the _Content Management_ vernacular. However, perhaps I need/want to intercept the _part_ instance itself and configure its corresponding _assembly(ies)_, _assembly name(s)_, and so forth, if at all possible. Probably as an _extension method_ during `Startup` _config_, _services_, or whatever. Not sure if that is possible?

**Reply by mwpowellhtx** [reply]
Perhaps the way to do that today is via `appsettings.json`, i.e. relay a _class name_, etc, via configuration for consumption by the module (?).

**Reply by mwpowellhtx** [reply]
Instead of that approach, could just have DI provide the part with an `IAssemblyCredentialInfo` instance, i.e.

```csharp
public interface IAssemblyCredentialInfo
{
  Assembly Assembly { get; }
}
```

Which can be wired during the _CMS_ project `Startup`.

**Reply by mwpowellhtx** [reply]
Wired that and fed an instance to the `ContentPartHandler<TPart>` class; although not real happy with the name. Too generic IMO; `EventHandlers` or even `Handlers` might be better and convey more meaning. Also comments! Or lack thereof; self-descriptive comments are a must. But I digress... through which I can relay the `Info` to the `part` when we work with the instance.

**Comment by Skrypt** [comment]
I'm currently not sure if what you are trying to achieve makes sense to be done through a ContentPart. Adding Assemblies version and descriptions are generally something that you will achieve by using a Properties.cs file inside a .csproj module. Here, we also have Manifest.cs files for describing `Orchard Core` specific modules details.

I think you should definitely look at Properties.cs files inside a .net core project (.csproj). That's how you generally set an assembly copyrights details and else.

**Reply by mwpowellhtx** [reply]
I have a [bump versions](https://nuget.org/packages/BumpAssemblyVersions/) that bumps the `.csproj` versions in situ, via the `XML` itself, done at the appropriate phase line during the build process itself. Following which, whatever conditions, dependencies, etc, should, ostensibly, respond accordingly, use the same variables, etc, I should think.

**Reply by mwpowellhtx** [reply]
To clarify, is not something that we generally set statically, let alone manually. We automate that during a build.

**Reply by Skrypt** [reply]
I remember using something similar. Let me think about it overnight. Maybe this is something that we could actually use in source code. Currently, we change manually the VersionPrefix in the `Properties.cs` of our modules. We have an `OrchardCore.Commons.props` file that applies the same details for each of our modules.

**Reply by mwpowellhtx** [reply]
For perspective, trying to assemble something not unlike this, via `OrchardCore` _CMS_ framework elements, and connect it with the _top level application assembly_ and such.

![image](https://user-images.githubusercontent.com/1824187/156850749-590f8286-22a8-499f-97e5-07483fddca7b.png)

**Reply by Skrypt** [reply]
I will read your module documentation. Need to find if that would work with what we currently have. Also, need to think about how we manage the versioning of our assemblies (but that's another story). 😄

**Reply by mwpowellhtx** [reply]
Well, again, that's how we like to manage our versioning. How you manage yours is up to you. 😉 Cheers, sir, thanks so much.

**Reply by mwpowellhtx** [reply]
> I will read your module documentation. Need to find if that would work with what we currently have. Also, need to think about how we manage the versioning of our assemblies (but that's another story). 😄

Bottom line is that it is evaluated during the build steps. As long as that happens, the normal versions that can be _available_, _version_, _package_, _assembly_, _file_, etc, all good. The variables are what they are, as long as that can be connected i.e. perhaps during a task depending on those phase lines, that should be fine.

**Reply by mwpowellhtx** [reply]
> For perspective, trying to assemble something not unlike this, via `OrchardCore` _CMS_ framework elements, and connect it with the _top level application assembly_ and such.

Did a bit more experimentation, as I think I mentioned herein (or this is the first if not second time), I can connect the Assembly details through DI factories to the handlers and relay that during the _handlers_ `Initializing` phase line. Perfect.

Also working with a 3P nuget package `LibGit2Sharp` for purposes of possibly informing about SHA or other repository details, i.e. during a debugging session. Has potential, whether part of the same supporting _'model'_ (loosely called) or resource.

Then I think it is a matter of the views that are involved. So... I think it is possible, cautiously optimistic at this point.

**Reply by mwpowellhtx** [reply]
> ...need to think about how we manage the versioning of our assemblies (but that's another story). 😄

Indeed. A bit more in depth thinking, @Skrypt. There has no doubt been adoption into the `ModuleAttribute` usage; but for versioning, I think the core attributes may already exist in one form or another via `Assembly`, `AssemblyName`, and so forth. Not sure why that does not work for y'all, and the constructs are well established in the .NET framework.

Also re: `csproj` properties, there's a whole set of them for versioning purposes, from simple `Version`, to `Assembly`, `File`, `Package`, and so forth. I support that via my `BAV` tooling. Which build turns around and generates the traditional `AssemblyInfo`, i.e. perhaps the `Properties` that you are thinking of, using that and some other elements. But you have to dig a little bit in the _Microsoft_ docs to get to much of that detail; and, like I stated, we like to automate that versioning as much as possible.

The tricky part for us, I think, is that `BAV` rewrites the project file, IIRC, been little while since I reviewed it. I think this is reloaded by the build process, but my memory is a bit fuzzy where that is concerned. So not sure whether we have an up to date property value to feed `ModuleAttribute`. Or direct support for the assembly version would be the best possible case, I think; but how deep is your adoption into `ModuleAttribute` at this point, I cannot answer that for you and the _Orchard_ team/community.

**Comment by mwpowellhtx** [comment]
[Side issue](https://github.com/OrchardCMS/OrchardCore/issues/11314) when editing with the custom part added to the content type.

**Comment by mwpowellhtx** [comment]
Discussed with @sebastienros during triage Thu, 31 Mar 2022. Will file a more specific issue and submit PR along these lines. Thanks...
